multicore_runtime/wsgi.py

```diff
@@ -0,0 +1,128 @@
+# Copyright 2015 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""Configure a user project and instantiate WSGI app meta_app to serve it.
+
+Importing this module will result in side-effects, such as registering the
+project's default ticket via vmstub.Register. This is a broadly compatible way
+to induce the webserver to run initialization code before starting to serve the
+WSGI app.
+"""
+
+import logging
+import os
+
+import cloud_logging
+from dispatcher import dispatcher
+from middleware import health_check_middleware
+from middleware import reset_environment_middleware
+from wsgi_config import env_vars_from_env_config
+from wsgi_config import get_module_config
+from wsgi_config import get_module_config_filename
+from wsgi_config import load_user_scripts_into_handlers
+from wsgi_config import ThreadLocalDict
+from wsgi_config import user_env_vars_from_appinfo
+
+from google.appengine.ext.vmruntime import vmconfig
+from google.appengine.ext.vmruntime import vmstub
+
+# Configure logging to output structured JSON to Cloud Logging.
+root_logger = logging.getLogger('')
+try:
+  handler = cloud_logging.CloudLoggingHandler()
+  root_logger.addHandler(handler)
+except IOError:
+  # If the Cloud Logging endpoint does not exist, just use the default handler
+  # instead. This will be the case when running in local dev mode.
+  pass
+
+root_logger.setLevel(logging.INFO)
+
+# Fetch application configuration via the config file.
+appinfo = get_module_config(get_module_config_filename())
+env_config = vmconfig.BuildVmAppengineEnvConfig()
+
+# Ensure API requests include a valid ticket by default.
+vmstub.Register(vmstub.VMStub(env_config.default_ticket))
+
+# Take an immutable snapshot of env data from env_config. This is added to the
+# environment in `reset_environment_middleware` in a particular order to ensure
+# that it cannot be overridden by other steps.
+frozen_env_config_env = tuple(
```

Comment:
As with the other mention of the `tuple()` question, because it is so critical the frozen env not be modified after creation time, I think it might be better to leave it as a tuple so that the env is immutable. I agree it looks awkward, but there's no convenient immutable sequence object in Python other than the tuple.